File: crypto_code/LAB3/AESServer.java
import java.net.*;
import java.io.*;
import java.util.Scanner;

public class AESServer {
    public static void main(String[] args) throws Exception {

        ServerSocket ss = new ServerSocket(8060);
        System.out.println("AES Server Started on port 8060...");
        System.out.println("Waiting for client connection...\n");

        while (true) {
            Socket s = ss.accept();
            System.out.println("Client connected!");

            BufferedReader in = new BufferedReader(
                    new InputStreamReader(s.getInputStream()));

            String ciphertext = in.readLine();

            if (ciphertext != null) {
                System.out.println("========== RECEIVED FROM CLIENT ==========");
                System.out.println("Received Ciphertext: " + ciphertext);
                System.out.println();

                Scanner sc = new Scanner(System.in);
                System.out.print("Enter 16-character Key for Decryption: ");
                String keyStr = sc.nextLine();
                while (keyStr.length() < 16) keyStr += "0";
                keyStr = keyStr.substring(0, 16);

                System.out.println("\n========== AES-128 DECRYPTION AT SERVER ==========\n");
                String decrypted = AESCore.decrypt(ciphertext, keyStr);

                String decryptedASCII = hexToASCII(decrypted);
                System.out.println("Decrypted Plaintext (Hex): " + decrypted);
                System.out.println("Decrypted Plaintext (ASCII): " + decryptedASCII);
            }

            s.close();
            break; // Close server after one client
        }

        ss.close();
    }

    private static String hexToASCII(String hex) {
        StringBuilder result = new StringBuilder();
        for (int i = 0; i < hex.length(); i += 2) {
            String hexByte = hex.substring(i, i + 2);
            int charCode = Integer.parseInt(hexByte, 16);
            result.append((char) charCode);
        }
        return result.toString();
    }
}

File: crypto_code/LAB3/AESServer.py
#!/usr/bin/env python3
"""
AES Server - Receives encrypted data and decrypts it
Port: 8060
"""

import socket
from AES import decrypt

HOST = '127.0.0.1'
PORT = 8060

# Create TCP socket
server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
server_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
server_socket.bind((HOST, PORT))

# Listen for incoming connections
server_socket.listen(1)
print(f"AES Server is listening on {HOST}:{PORT}...")

# Accept client connection
conn, addr = server_socket.accept()
print(f"\nConnected by {addr}\n")

try:
    # Receive ciphertext from client
    ciphertext = conn.recv(1024).decode().strip()
    print(f"========== RECEIVED FROM CLIENT ==========")
    print(f"Received Ciphertext: {ciphertext}\n")
    
    # Get key from user
    key_str = input("Enter 16-character Key for Decryption: ").strip()
    
    # Pad or truncate key to 16 characters
    key_str = (key_str + "0" * 16)[:16]
    print(f"Key: {key_str}\n")
    
    # Decrypt
    print("========== AES-128 DECRYPTION AT SERVER ==========")
    decrypted = decrypt(ciphertext, key_str)
    
    # Convert hex to ASCII using built-in bytes.fromhex()
    decrypted_ascii = bytes.fromhex(decrypted.replace(" ", "").replace("\n", "")).decode('utf-8', errors='replace')
    
    print(f"\n========== DECRYPTION RESULT ==========")
    print(f"Decrypted Plaintext (ASCII): {decrypted_ascii}")
    print(f"Decrypted Plaintext (Hex): {decrypted}\n")
    
    # Send result back to client
    conn.sendall(decrypted_ascii.encode())
    print(f"Sent decrypted plaintext back to client.")
    
except Exception as e:
    print(f"Error: {e}")
finally:
    conn.close()
    server_socket.close()
    print("\nServer closed.")
